feat: ilp-cli tool for interacting with nodes (#334)
interledger-rs/interledger-rs: examples/simple/README.md

### 3. Launch 2 Nodes

<!--!
printf "\n\nStarting Interledger nodes...\n"
if [ "$USE_DOCKER" -eq 1 ]; then
    $CMD_DOCKER run \
        -e ILP_ADDRESS=example.node_a \
        -e ILP_SECRET_SEED=8852500887504328225458511465394229327394647958135038836332350604 \

>         -e ILP_ADMIN_AUTH_TOKEN=admin-example \

Reviewer: we had admin-a and admin-b to show that each node can use a different auth token, think it'd be better to leave that as is

Author: I believe that's evident from the `interledger node` command itself. Let's centralize discussion in https://github.com/interledger-rs/interledger-rs/pull/334#discussion_r328406029 .

wait_to_serve "http://localhost:7770" 10 || error_and_exit "\nFailed to spin up nodes. Check out your configuration and log files."
wait_to_serve "http://localhost:8770" 10 || error_and_exit "\nFailed to spin up nodes. Check out your configuration and log files."

> printf " done\nThe Interledger.rs nodes are up and running!\n\n"

Reviewer 2: ```suggestion
printf "done\nThe Interledger.rs nodes are up and running!\n\n"
```

Author: This space is deliberate because the preceding output is the line of repeated `.` that we print while waiting for the nodes to spin up, so this makes the difference between the output looking like `Waiting for Interledger.rs nodes to start up.....done` and `Waiting for Interledger.rs nodes to start up..... done`.

See the [HTTP API docs](../../docs/api.md) for the full list of fields that can be set on an account.

<!--!
printf "\nCreating accounts...\n\n"

if [ "$USE_DOCKER" -eq 1 ]; then
    printf "Alice's account:\n"
    curl \
        -H "Content-Type: application/json" \

>         -H "Authorization: Bearer admin-example" \

Reviewer 2: Why is this being changed? This is meant to show it using the admin credentials for different nodes

Author: I'm under the impression that we consider these examples unrepresentative because we don't actually expect anyone to host two logical nodes on the same machine. Furthermore, for any extensive use of this CLI I would encourage people to use the environment variable to avoid the need to type their password repeatedly. In that sense using different passwords for each logical node seems irrelevant to the example, and requiring that the example include additional command-line flags only to accommodate an atypical use case would seem to be a distraction from what we expect the user to take away from this example.

Author: Separate passwords reinstated; now that I've also begin using a symbolic link to get of the `cargo run` invocation, that leaves plenty of room for the occasional `--auth admin-b` argument that the one node will need, and manages to nicely show off the features of the CLI at the same time.

```bash
# Insert accounts on Node A
# One account represents Alice and the other represents Node B's account with Node A

> export ILP_CLI_AUTH=admin-example

Reviewer 2: I think we can either re-export this before specific calls or pass this in as a CLI arg in each call so that we don't need to make all the nodes use the same admin auth token

Author: Let's centralize this discussion at https://github.com/interledger-rs/interledger-rs/pull/334#discussion_r328406029.

printf "Creating Alice's account on Node A...\n"

> cargo run --quiet --bin ilp-cli -- \

Reviewer 2: Ugh this command does not look amazing (because of the `cargo run...` part in front). Would it be terrible to do something like aliasing `ilp-cli` to that whole string? If you were really going to install the CLI on your system you'd do `cargo install ilp-cli` so that it would be available in your path without needing `cargo run...`

Reviewer: I think we can just add a step in the top level of the examples that they should do `cargo install --release X` and replace all occurrences of `cargo run` with the binary name (or with `./target/{debug,release}/X`

Author: > I think we can just add a step in the top level of the examples that they should do cargo install --release X and replace all occurrences of cargo run with the binary name (or with ./target/{debug,release}/X

I'll explore the former, but with regard to the latter we might have trouble using a relative path because we can't guarantee from which directory the user will be running this example; we could require that they be in `examples/simple` if that seems reasonable.

Author: In the interest of expedience I'm changing these invocations to just `../../target/debug/ilp-cli`, which could be nicer with an alias or symbolic link, but is a good compromise for the moment and is the simplest and most transparent thing that works.

Reviewer 2: Maybe it's not so bad. Do we need to run it with `--quiet`?

Author: We do want that flag because otherwise you get a few lines of bright green cargo output after every single invocation. I've actually gone even further and begun using a symbolic link here, which ends up looking very nice.

-->

```bash

> # We merely use this symbolic link here to pretend that ilp-cli is installed

Reviewer 2: ```suggestion
# We merely use this symbolic link here to pretend that ilp-cli is installed
# You would not need this line if you installed `ilp-cli` by running `cargo install ilp-cli`
```

Author: My only concern with suggesting that people use `cargo install` is that I worry about a situation where someone has a stale version of the CLI installed which might end up causing errors that are frustrating for users to diagnose down the road. Let me know if you think the risk here isn't worth worrying about.

Reviewer 2: Not worth it in this case. If the CLI isn't stable right now, it should be very soon

ln -s ../../target/debug/ilp-cli ilp-cli

> # For authenticating to nodes, we can set credentials as follows

Reviewer 2: ```suggestion
# For authenticating to nodes, we can set credentials as an environment variable or a CLI argument
```

-->

```bash
# We merely use this symbolic link here to pretend that ilp-cli is installed
ln -s ../../target/debug/ilp-cli ilp-cli

# For authenticating to nodes, we can set credentials as an environment variable or a CLI argument
export ILP_CLI_API_AUTH=admin-a

# Insert accounts on Node A
# One account represents Alice and the other represents Node B's account with Node A

printf "Creating Alice's account on Node A...\n"

> ./ilp-cli --quiet accounts create alice \

Reviewer 2: Beautiful

# For integration tests
function hook_before_kill() {
    if [ $TEST_MODE -eq 1 ]; then

>         test_equals_or_exit '{"balance":"-500"}' test_http_response_body -H "Authorization: Bearer admin-example" http://localhost:7770/accounts/alice/balance

Reviewer 2: Should the tests also use the CLI? @bstrie @dora-gt

Reviewer: Yeah I'd prefer that

Reviewer 3: hmmmm difficult question.

This simply checks, finally it worked well or not. So using `cli` here means mixing adulterants in. That said, actually now this is working as an integration test, using `cli` makes it possible to test `cli` itself. It might be better.

Author: I do like the idea of using this to test the CLI as well; since this is a monorepo I think it's reasonable to make the CI test the examples and the CLI in the same test.
